### PR TITLE
fix(neighbors): detect cluster-tile buffer overflow

### LINKED ISSUE #159: `NeighborOverflowError` has different semantics when `cluster_tile` is selected, breaking the documented "increase max_neighbors" contract
## Environment

- nvalchemi-toolkit-ops 0.4.1
- CUDA GPU

## Description

`NeighborOverflowError`'s docstring states:

> Users should increase the relevant ``max_neighbors`` / segment-capacity parameter or provide a larger pre-allocated tensor.

This contract holds for the naive and cell-list methods, where the exception means a
per-atom neighbor count exceeded `max_neighbors`. However, when `cluster_tile` is
selected, the same exception type is raised for a different
condition: overflow of the internal tile-pair buffer. That buffer's capacity is
estimated from atom count, cutoff, and cell volume, and does **not** depend on
`max_neighbors` — and the public `neighbor_list()` API has no documented parameter to
grow it.

So for a caller hitting this case:

1. The documented remedy (increase `max_neighbors`) does not work.
2. There is no way to tell the two conditions apart: the exception type, fields, and message all look like a `max_neighbors` overflow.

## Suggestion

Would you consider raising a distinguishable error for the tile-buffer overflow case?
Alternatively, documenting the differing semantics or handling the tile-buffer growth
internally would also resolve this.

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Detect exhausted intermediate cluster-tile buffers in eager Torch and JAX neighbor-list paths, report the required capacity with a dedicated exception, and define explicit capacity requirements for transformed JAX calls. Document allocation, retry sizing, and compiled post-call validation.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #159

## Changes Made

- Add `TileBufferOverflow` with required count, allocated capacity, and segmented-system context, and use it across eager Torch and JAX cluster-tile output paths.
- Add explicit `max_tiles_per_group` controls to cluster-tile bindings and require a positive static Python integer for transformed or compiled JAX calls.
- Extend existing overflow tests and document heuristic estimation, exact eager retry sizing, worst-case bounds, and compiled count validation.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

None.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Review by laserkelvin — COMMENTED** [review]
I found one blocking direct-builder contract gap and two follow-up test/validation gaps. Details are inline.

File: docs/userguide/about/migration.md [hunk 1]
```diff
@@ -8,6 +8,31 @@ This guide lists user-visible migrations by release.
 
 ## Unreleased
 
+### Cluster-Tile Buffer Capacity
+
+Cluster-tile construction writes candidate tile pairs to a fixed-size
+intermediate buffer before producing a neighbor matrix or COO list. Previously,
+some eager return paths did not check whether construction filled that buffer,
+while existing guards reported the condition as `NeighborOverflowError`. All
+eager Torch and JAX cluster-tile paths now raise `TileBufferOverflow` when the
```
Comment (laserkelvin) [inline docs/userguide/about/migration.md:17]: **blocking:** The one-shot JAX wrapper checks this count before querying, but the public lower-level builders do not. Their Warp build increments `num_tiles` for every candidate and writes only while the fixed buffer has room. That leaves direct eager callers with a valid-looking state whose tile arrays contain only a prefix:

```python
state = build_cluster_tile_list(..., max_tiles_per_group=1)
query_cluster_tile(..., state[11], state[12], state[13], ...)
```

The query then returns an incomplete neighbor list unless the caller manually compares the returned count with the buffer capacity. The migration guide says all eager JAX paths raise, so please call `_check_eager_tile_buffer_capacity()` in the eager builders too (and their batch sibling), or explicitly narrow that contract.

File: nvalchemiops/jax/neighbors/cluster_tile.py [hunk 2]
```diff
@@ -102,27 +102,39 @@ def _concrete_cell_volume(cell) -> float | None:
...
-    more memory for very large traced systems).
+    """Choose ``max_tiles_per_group`` for a JAX compact tile buffer.
+
+    ``max_tiles_per_group`` determines an array shape and must therefore be a
+    static Python integer. Concrete eager inputs use the geometry estimator.
+    When an input is traced, runtime geometry is unavailable and the caller
+    must provide the value explicitly.
     """
     ngroup = (int(total_atoms) + TILE_GROUP_SIZE - 1) // TILE_GROUP_SIZE
     if ngroup <= 1:
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/cluster_tile.py:122]: **suggestion:** This works today because a one-group system has a fixed capacity of one. It still bypasses the stated transformed-JAX rule: `max_tiles_per_group` must be an explicit static Python integer because it determines output shape. Move the traced-input validation before this early return, then add a `jax.jit` regression using <=32 atoms and `max_tiles_per_group=None` that raises the documented `ValueError`.

File: test/neighbors/bindings/jax/test_cluster_tile.py [hunk 3]
```diff
@@ -420,6 +424,24 @@ def test_invalid_format_raises(self):
...
+            with pytest.raises(TileBufferOverflow) as caught:
+                cluster_tile_neighbor_list(
+                    positions,
+                    5.0,
+                    cell,
+                    max_neighbors=256,
+                    max_tiles_per_group=1,
+                    return_distances=return_distances,
+                )
+            assert caught.value.num_tiles > caught.value.max_tiles
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_cluster_tile.py:444]: **suggestion:** These checks show that the matrix convenience API raises, but they do not protect the full overflow contract. Please add direct eager build -> query coverage: build with a deliberately small bound, verify that it raises before query, and compare the exception's `num_tiles` with a sufficiently sized build so it proves the retry size is correct. The Torch COO and tile-format exits should receive the same coverage.

File: docs/userguide/about/migration.md (hunk 1 above)
Comment (zubatyuk) [reply docs/userguide/about/migration.md:17]: Addressed in 8547f768

File: nvalchemiops/jax/neighbors/cluster_tile.py (hunk 2 above)
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/cluster_tile.py:122]: Addressed in 8547f768

File: test/neighbors/bindings/jax/test_cluster_tile.py (hunk 3 above)
Comment (zubatyuk) [reply test/neighbors/bindings/jax/test_cluster_tile.py:444]: Addressed in 8547f768

**Comment by dallasfoster** [comment]
/ok to test [1b3049d](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/164/commits/1b3049df427a195e82963a612f0a9c1e09191947)

**Review by dallasfoster — APPROVED** [review]
